pkg/lib/machine-dialogs.js

```diff
@@ -607,68 +667,151 @@ function HostKey(dialog, problem) {
+            return;
 
-            for (const id in keys.items) {
-                const key = keys.items[id];
-                if (key.name === identity_path.split("/").pop() && key.loaded)
-                    dialog.get_sel(".login-locked").empty();
-            }
+        old_extra_state = state;
+        $("#login-setup-text").text(text);
+        $("#login-setup-extra").empty()
+                .append(extra);
+    }
+
+    function update_auth() {
+        if (offer_login_password && offer_key_password) {
+            dialog.get_sel("#login-authentication, #login-authentication + *").show();
+            use_login_password = dialog.get_sel("#login-authentication + div input[value=pass]").prop('checked');
+        } else {
+            dialog.get_sel("#login-authentication, #login-authentication + *").hide();
+            use_login_password = offer_login_password;
         }
+        use_key_password = offer_key_password && !use_login_password;
+
+        dialog.get_sel("#login-diff-password, #login-diff-password + *")
+                .toggle(use_login_password);
 
-        var empty_div = dialog.get_sel('.empty');
-        if (empty_div)
-            empty_div.toggleClass("hidden", have_keys);
+        dialog.get_sel(".login-locked, .login-locked + *, .login-locked + * + *")
+                .toggle(use_key_password);
+
+        if (!default_ssh_key)
+            offer_key_setup = false;
+        else if (default_ssh_key.unaligned_passphrase)
+            offer_key_setup = use_key_password;
+        else
+            offer_key_setup = true;
     }
 
-    function login() {
-        var address;
-        var options = {};
-        var dfp = $.Deferred();
-        var user = $("#login-custom-user").val();
+    function toggle_setup_extra() {
+        dialog.get_sel(".login-setup-auto, .login-setup-auto + *").toggle(offer_key_setup);
+        $("#login-setup-extra").toggle(offer_key_setup && $('#login-setup-keys').prop('checked'));
+    }
 
-        var parts = dialog.machines_ins.split_connection_string(dialog.address);
-        parts.user = user;
+    function update() {
+        update_key_setup();
+        update_auth();
+        toggle_setup_extra();
+    }
 
-        address = dialog.machines_ins.generate_connection_string(parts.user,
-                                                                 parts.port,
-                                                                 parts.address);
+    function show_error(message, target) {
+        const ex = new Error(message);
+        ex.target = target;
+        dialog.get_sel().dialog("failure", ex);
+    }
 
-        if ($("#login-type button").val() != 'stored') {
+    function setup_keys(host, cur_passphrase, login_password) {
+        return cockpit.script(ssh_setup_default_key_sh, [default_ssh_key.name, cur_passphrase, login_password])
+                .then(function (data) {
+                    return cockpit.script(ssh_add_key_sh, [data.trim()], { host: host });
```

Comment:
Please use `.then(data => cockpit.script())`... arrow notation for brevity.

Moreover, please add a .catch(), in case the shell script fails. console.error() is probably enough, but let's avoid an entire page crash.

Reply:
Arrows done.  The error should end up in the dialog, no?

Reply:
I have included stderr from the scripts in the error message as well.

Reply:
I see, a promise rejection gets handled in the dialog.run() call.